File: examples/node-react/src/__tests__/notes.test.ts
/**
 * Domain tests. No render, no DOM, no act().
 *
 * These run in milliseconds because they touch nothing but plain objects. That
 * speed is why the publish rule lives outside the component.
 */
import { describe, expect, it } from "vitest";

import {
  MAX_BODY_LENGTH,
  type Note,
  NotFoundError,
  ValidationError,
  createNote,
  publishInList,
  publishNote,
} from "../notes";

const FIXED_NOW = "2026-01-02T03:04:05.000Z";
const LATER = "2026-06-07T08:09:10.000Z";

function draft(overrides: Partial<Note> = {}): Note {
  return {
    id: "n1",
    title: "Standup",
    body: "we shipped the gate runner",
    status: "draft",
    publishedAt: null,
    ...overrides,
  };
}

describe("createNote", () => {
  it("starts a note as a draft", () => {
    const note = createNote("n1", "Standup", "content");
    expect(note.status).toBe("draft");
    expect(note.publishedAt).toBeNull();
  });

  it("trims the title", () => {
    expect(createNote("n1", "  Standup  ", "content").title).toBe("Standup");
  });

  it("rejects an empty title", () => {
    expect(() => createNote("n1", "   ", "content")).toThrow(ValidationError);
  });

  it("rejects an oversized body", () => {
    const tooLong = "x".repeat(MAX_BODY_LENGTH + 1);
    expect(() => createNote("n1", "Standup", tooLong)).toThrowError(/at most/);
  });
});

describe("publishNote", () => {
  it("sets status and timestamp on the happy path", () => {
    const note = publishNote(draft(), () => FIXED_NOW);
    expect(note.status).toBe("published");
    expect(note.publishedAt).toBe(FIXED_NOW);
  });

  it("does not mutate the input note", () => {
    const original = draft();
    publishNote(original, () => FIXED_NOW);
    expect(original.status).toBe("draft");
    expect(original.publishedAt).toBeNull();
  });

  it("rejects an empty body", () => {
    expect(() => publishNote(draft({ body: "" }), () => FIXED_NOW)).toThrow(ValidationError);
  });

  it("rejects a whitespace only body", () => {
    // The edge case a naive `if (!body)` check gets right and a naive
    // `if (body === undefined)` check gets wrong.
    expect(() => publishNote(draft({ body: "  \n\t " }), () => FIXED_NOW)).toThrow(ValidationError);
  });

  it("is idempotent and keeps the first timestamp", () => {
    // The edge case. The clock returns a DIFFERENT time on the second call, so a
    // bug that re-stamps publishedAt fails this test loudly.
    const first = publishNote(draft(), () => FIXED_NOW);
    const second = publishNote(first, () => LATER);

    expect(second.status).toBe("published");
    expect(second.publishedAt).toBe(FIXED_NOW);
    // Same reference, so React skips the re-render.
    expect(second).toBe(first);
  });

  it("keeps an already published note published even if the body was blanked", () => {
    // Guards the rule order inside publishNote. Swapping the two checks makes
    // this go red, which is what proves the order is deliberate.
    const published = publishNote(draft(), () => FIXED_NOW);
    const blanked: Note = { ...published, body: "" };

    const result = publishNote(blanked, () => LATER);
    expect(result.status).toBe("published");
    expect(result.publishedAt).toBe(FIXED_NOW);
  });
});

describe("publishInList", () => {
  it("publishes the matching note and leaves the others alone", () => {
    const notes = [draft({ id: "n1" }), draft({ id: "n2" })];
    const result = publishInList(notes, "n2", () => FIXED_NOW);

    expect(result[0]).toBe(notes[0]);
    expect(result[1].status).toBe("published");
  });

  it("returns the SAME array when the note was already published", () => {
    // This is the render-skip contract. A new array every call would re-render
    // the whole list on every retry.
    const notes = [publishNote(draft(), () => FIXED_NOW)];
    expect(publishInList(notes, "n1", () => LATER)).toBe(notes);
  });

  it("throws NotFoundError for an unknown id", () => {
    expect(() => publishInList([], "nope", () => FIXED_NOW)).toThrow(NotFoundError);
  });
});

File: examples/go/internal/notes/notes_test.go
package notes

import (
	"errors"
	"strings"
	"sync"
	"testing"
	"time"
)

var (
	fixedNow = time.Date(2026, 1, 2, 3, 4, 5, 0, time.UTC)
	later    = time.Date(2026, 6, 7, 8, 9, 10, 0, time.UTC)
)

// clockAt returns a clock that hands out times from a fixed list, one per call.
// A test can then tell the first publish apart from the second.
func clockAt(times ...time.Time) func() time.Time {
	var i int
	var mu sync.Mutex
	return func() time.Time {
		mu.Lock()
		defer mu.Unlock()
		t := times[len(times)-1]
		if i < len(times) {
			t = times[i]
			i++
		}
		return t
	}
}

func TestCreateStartsAsDraft(t *testing.T) {
	t.Parallel()
	s := NewStoreWithClock(clockAt(fixedNow))

	note, err := s.Create("Standup", "we shipped the gate runner")
	if err != nil {
		t.Fatalf("Create returned error: %v", err)
	}
	if note.Status != StatusDraft {
		t.Errorf("status = %q, want %q", note.Status, StatusDraft)
	}
	if note.PublishedAt != nil {
		t.Errorf("PublishedAt = %v, want nil", note.PublishedAt)
	}
}

func TestCreateValidation(t *testing.T) {
	t.Parallel()

	cases := []struct {
		name      string
		title     string
		body      string
		wantField string
	}{
		{name: "empty title", title: "", body: "x", wantField: "title"},
		{name: "whitespace title", title: "   ", body: "x", wantField: "title"},
		{name: "oversized body", title: "ok", body: strings.Repeat("x", MaxBodyLength+1), wantField: "body"},
	}

	for _, tc := range cases {
		// Go 1.22 gave every loop iteration its own variable, so the old
		// `tc := tc` line is no longer needed. See the Dev agent record in
		// specs/epic-01-notes/story-01-publish-note.md: on Go 1.21 and earlier,
		// t.Parallel() inside a subtest made every case read the LAST tc, and
		// the table still passed because the last case happened to be valid.
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			s := NewStoreWithClock(clockAt(fixedNow))

			_, err := s.Create(tc.title, tc.body)
			var ve *ValidationError
			if !errors.As(err, &ve) {
				t.Fatalf("Create(%q, len=%d) error = %v, want ValidationError", tc.title, len(tc.body), err)
			}
			if ve.Field != tc.wantField {
				t.Errorf("field = %q, want %q", ve.Field, tc.wantField)
			}
		})
	}
}

func TestPublishHappyPath(t *testing.T) {
	t.Parallel()
	s := NewStoreWithClock(clockAt(fixedNow))
	note, _ := s.Create("Standup", "we shipped the gate runner")

	got, err := s.Publish(note.ID)
	if err != nil {
		t.Fatalf("Publish returned error: %v", err)
	}
	if got.Status != StatusPublished {
		t.Errorf("status = %q, want %q", got.Status, StatusPublished)
	}
	if got.PublishedAt == nil || !got.PublishedAt.Equal(fixedNow) {
		t.Errorf("PublishedAt = %v, want %v", got.PublishedAt, fixedNow)
	}
}

func TestPublishRejectsBlankBody(t *testing.T) {
	t.Parallel()

	cases := []struct{ name, body string }{
		{name: "empty", body: ""},
		{name: "whitespace only", body: "  \n\t "},
	}

	for _, tc := range cases {
		t.Run(tc.name, func(t *testing.T) {
			t.Parallel()
			s := NewStoreWithClock(clockAt(fixedNow))
			note, _ := s.Create("Standup", tc.body)

			_, err := s.Publish(note.ID)
			var ve *ValidationError
			if !errors.As(err, &ve) {
				t.Fatalf("Publish error = %v, want ValidationError", err)
			}
			if ve.Field != "body" {
				t.Errorf("field = %q, want body", ve.Field)
			}
			stored, _ := s.Get(note.ID)
			if stored.Status != StatusDraft {
				t.Errorf("status = %q, want it to stay %q", stored.Status, StatusDraft)
			}
		})
	}
}

// TestPublishIsIdempotent is the edge case. A second publish is not an error and
// does not move the timestamp. The clock hands out `later` on the second call,
// so a bug that re-stamps PublishedAt fails this test loudly.
func TestPublishIsIdempotent(t *testing.T) {
	t.Parallel()
	s := NewStoreWithClock(clockAt(fixedNow, later))
	note, _ := s.Create("Standup", "we shipped the gate runner")

	first, err := s.Publish(note.ID)
	if err != nil {
		t.Fatalf("first Publish returned error: %v", err)
	}
	second, err := s.Publish(note.ID)
	if err != nil {
		t.Fatalf("second Publish returned error: %v, want nil", err)
	}

	if second.Status != StatusPublished {
		t.Errorf("status = %q, want %q", second.Status, StatusPublished)
	}
	if !second.PublishedAt.Equal(*first.PublishedAt) {
		t.Errorf("second publish moved PublishedAt from %v to %v", first.PublishedAt, second.PublishedAt)
	}
	if !second.PublishedAt.Equal(fixedNow) {
		t.Errorf("PublishedAt = %v, want the first publish time %v", second.PublishedAt, fixedNow)
	}
}

// Guards the rule order inside Publish. Once published, the note stays published
// even if the body was later emptied. Swapping the two checks makes this go red.
func TestPublishOfPublishedNoteSurvivesBlankedBody(t *testing.T) {
	t.Parallel()
	s := NewStoreWithClock(clockAt(fixedNow, later))
	note, _ := s.Create("Standup", "content")
	if _, err := s.Publish(note.ID); err != nil {
		t.Fatalf("first Publish returned error: %v", err)
	}

	stored, _ := s.Get(note.ID)
	stored.Body = ""

	got, err := s.Publish(note.ID)
	if err != nil {
		t.Fatalf("Publish after blanking body returned error: %v, want nil", err)
	}
	if got.Status != StatusPublished || !got.PublishedAt.Equal(fixedNow) {
		t.Errorf("got status %q at %v, want %q at %v", got.Status, got.PublishedAt, StatusPublished, fixedNow)
	}
}

func TestPublishUnknownIDIsNotFound(t *testing.T) {
	t.Parallel()
	s := NewStoreWithClock(clockAt(fixedNow))

	if _, err := s.Publish("does-not-exist"); !errors.Is(err, ErrNotFound) {
		t.Fatalf("error = %v, want ErrNotFound", err)
	}
}

// Concurrent publishes must produce exactly one timestamp. Run this under
// `go test -race` to make the mutex do real work.
func TestConcurrentPublishStampsOnce(t *testing.T) {
	t.Parallel()
	s := NewStoreWithClock(clockAt(fixedNow, later, later, later, later))
	note, _ := s.Create("Standup", "content")

	var wg sync.WaitGroup
	for range 8 {
		wg.Add(1)
		go func() {
			defer wg.Done()
			_, _ = s.Publish(note.ID)
		}()
	}
	wg.Wait()

	stored, _ := s.Get(note.ID)
	if !stored.PublishedAt.Equal(fixedNow) {
		t.Errorf("PublishedAt = %v, want the first publish time %v", stored.PublishedAt, fixedNow)
	}
}
